### PR TITLE
Smart contract refactor and match test refactor

### PR DESCRIPTION
**Pull request by PaulZhemanov**
- **new main tests**
- **code refactor**
- **test_utils upgrade**
- **rename in tests**
- **little code refactor**
- **added review coments**
- **some fixes**
- **info abi added**
- **rever test added**
- **should panic onboard**
- **abi refactor**
- **toolchain config added**
- **gitignore refactor**
- **tests utils refactor**
- **some errors deleted**
- **fmt**
- **init_wallets init_token init_orderbook added**
- **setup added**

### CODE REVIEW

**Review by Braqzen — COMMENTED** [review]
I've glanced through this and there's numerous suggestions that I would make so I'm happy to just approve and do them all myself in the PR I'm working on after this gets merged and I have time

**Review by chlenc — APPROVED** [review]

**Comment by Braqzen** [comment]
PR is open but changes are being made. Is this ready to be merged?
I'm waiting to pull the changes. I'd rather not resolve merge conflicts.

**Comment by PaulZhemanov** [comment]
> PR is open but changes are being made. Is this ready to be merged? I'm waiting to pull the changes. I'd rather not resolve merge conflicts.

Let's work in the match-refactor branch. I got mistake in match_orders test.

File: contract/src/main.sw [hunk 1]
```diff
@@ -84,22 +84,17 @@ impl OrderBook for Contract {
...
-            size,
-            price,
             PRICE_DECIMALS,
             QUOTE_TOKEN_DECIMALS,
             QUOTE_TOKEN,
         );
 
-        require(msg_amount() == amount, AssetError::InvalidAssetAmount);
-        require(msg_asset_id() == asset, AssetError::InvalidAsset);
+        require(msg_amount() == asset.amount, AssetError::InvalidAssetAmount);
```
Comment (PaulZhemanov) [inline contract/src/main.sw:96]: @Braqzen can we check this at the beginning of the function?

**Pull request closed without merging**